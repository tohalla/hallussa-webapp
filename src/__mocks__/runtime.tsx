import { render as tlRender, RenderOptions } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";

import store from "../store/store";

export const render = jest.fn((ui: React.ReactElement, options?: Omit<RenderOptions, "queries">) =>
  tlRender(
    <Provider store={store}>{ui}</Provider>,
    options
  )
);
