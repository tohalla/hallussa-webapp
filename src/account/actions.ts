import { compose, find, head, prop } from "ramda";
import { AnyAction, Dispatch } from "redux";

import { CALL_API } from "../store/middleware/api/actions";
import { ReduxAPICall } from "../store/middleware/api/api";

export const SET_ACTIVE_ACCOUNT = "SET_ACTIVE_ACCOUNT";
export const FETCH_ACCOUNT_SUCCESS = "FETCH_ACCOUNT_SUCCESS";
export const FETCH_ACCOUNTS_SUCCESS = "FETCH_ACCOUNTS_SUCCESS";

export const ADD_ACCOUNT_SUCCESS = "ADD_ACCOUNT_SUCCESS";

export const SET_ACCOUNT_USER_ROLE = "SET_ACCOUNT_USER_ROLE";

export interface AccountPayload {
  id: number;
  firstName: string;
  lastName: string;
  email: string;
  organisations: ReadonlyArray<{userRole: number, organisation: number}>;
  userRole?: number;
}

export interface AccountAction extends AnyAction {
  type: string;
  payload?: Partial<AccountPayload>;
}

export const setActiveAccount = (payload: any) => ({
  payload,
  type: SET_ACTIVE_ACCOUNT,
});

export const fetchCurrentAccount = () => (dispatch: Dispatch) => dispatch<ReduxAPICall>({
  endpoint: "/accounts",
  method: "get",
  onSuccess: compose(dispatch, setActiveAccount, head, Object.keys as any),
  parameters: {eager: "organisations"},
  successType: FETCH_ACCOUNT_SUCCESS,
  transformResponse: (response) => ({[response.id]: response}),
  type: CALL_API,
});

export const fetchAccounts = (
  organisation: number,
  {bypassCache = false, accounts}: {bypassCache?: boolean, accounts?: number[]} = {}
): ReduxAPICall => ({
  endpoint: `/organisations/${organisation}/users/accounts${accounts ? `?accounts=[${
    accounts.join(",")
  }]` : ""}`,
  method: "get",
  successType: FETCH_ACCOUNTS_SUCCESS,
  type: CALL_API,
});

export const addAccount = (
  organisation: number,
  payload: {email: string, userRole?: number}
) => (dispatch: Dispatch) => dispatch<ReduxAPICall>({
  body: payload,
  endpoint: `/organisations/${organisation}/users/accounts`,
  method: "post",
  onSuccess: (responsePayload) => {
    const account = prop("account", responsePayload);
    if (typeof account === "number") {
      return dispatch(fetchAccounts(organisation, {accounts: [account]}));
    }
  },
  successType: ADD_ACCOUNT_SUCCESS,
  type: CALL_API,
});

export const setUserRole = (
  organisation: number,
  account: number,
  payload: {userRole?: number}
) => ({
  body: payload,
  endpoint: `/organisations/${organisation}/users/accounts/${account}`,
  method: "put",
  successType: SET_ACCOUNT_USER_ROLE,
  type: CALL_API,
});
