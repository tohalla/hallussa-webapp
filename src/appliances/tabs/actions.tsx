import {
  CHANGE_TAB_TO,
  CLOSE_ACTIVE_TAB,
  CREATE_NEW_TAB,
} from "./actionTypes";

export const addTab = () => ({
  // TODO: Fix payload
  type: CREATE_NEW_TAB,
});

export const closeTab = (identifier: number) => ({
  // TODO: Fix payload
  payload: {
    id: identifier,
  },
  type: CLOSE_ACTIVE_TAB,
});

export const changeTab = (identifier: number) => ({
  // TODO: Fix payload
  payload: {
    id: identifier,
  },
  type: CHANGE_TAB_TO,
});
