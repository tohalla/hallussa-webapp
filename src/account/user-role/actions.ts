import { find } from "ramda";

import { AnyAction } from "redux";
import { ThunkAction } from "redux-thunk";
import { Omit } from "../../../misc";
import { CALL_API } from "../../store/middleware/api/actions";
import { ReduxAPICall } from "../../store/middleware/api/api";
import { ReduxState } from "../../store/store";

export const FETCH_USER_ROLES_SUCCESS = "FETCH_USER_ROLES_SUCCESS";
export const SET_ACTIVE_USER_ROLE = "SET_ACTIVE_USER_ROLE";

export interface UserRolePayloadÂ {
  id: number;
  name: string;
  organisation: number;
  allowCreateAppliance: boolean;
  allowCreateMaintainer: boolean;
  allowDeleteAppliance: boolean;
  allowDeleteMaintainer: boolean;
  allowDeleteOrganisation: boolean;
  allowManageMaintenanceTask: boolean;
  allowManageRoles: boolean;
  allowManageUsers: boolean;
  allowUpdateAppliance: boolean;
  allowUpdateMaintainer: boolean;
  allowUpdateOrganisation: boolean;
  isShared: boolean;
}

export type UserRoleRights = Omit<
  UserRolePayload,
  "id" | "name" | "isShared" | "organisation"
> | {[key: string]: false};

export interface UserRoleAction {
  type: string;
  payload: UserRolePayload;
}

export const fetchRoles = (organisation?: number, {bypassCache = false} = {}): ReduxAPICall => ({
  endpoint: typeof organisation === "undefined" ? "/user-roles" : `/organisations/${organisation}/user-roles`,
  method: "get",
  successType: FETCH_USER_ROLES_SUCCESS,
  type: CALL_API,
});

export const updateActiveUserRole: ThunkAction<any, ReduxState, any, AnyAction> = (dispatch, getState) => {
  const {entities: {organisations}, session: {activeAccount, activeOrganisation}} = getState();
  if (activeAccount && activeOrganisation) {
    const joinRelation = find(
      ({account}) => String(account) === String(activeAccount),
      organisations[activeOrganisation].accounts || []
    );
    if (joinRelation) {
      return dispatch({
        payload: joinRelation.userRole,
        type: SET_ACTIVE_USER_ROLE,
      });
    }
  }

  return dispatch({
    type: SET_ACTIVE_USER_ROLE,
  });
};
