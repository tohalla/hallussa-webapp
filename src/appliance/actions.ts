import { find } from "ramda";
import { CALL_API } from "../store/middleware/api/actions";
import { ReduxAPICall } from "../store/middleware/api/api";

export const FETCH_APPLIANCES_SUCCESS = "FETCH_APPLIANCES_SUCCESS";

export interface AppliancePayloadÂ {
  id: number;
  hash: string;
  name: string;
  createdAt: string;
  updatedAt: string;
  maintainers: ReadonlyArray<number>;
}

export interface ApplianceAction {
  type: string;
  payload: AppliancePayload;
}

export const fetchAppliances = (organisation: number, {bypassCache = false} = {}): ReduxAPICall => ({
  attemptToFetchFromStore: bypassCache ? undefined : (state) =>
    !Boolean(find(// check if store contains all maintainers defined in organisation
      (appliance) => typeof state.entities.appliances[appliance] === "undefined",
      state.entities.organisations[organisation].appliances
    )) && state.entities.appliances,
  endpoint: `/organisations/${organisation}/appliances`,
  method: "get",
  parameters: {eager: "maintainers"},
  successType: FETCH_APPLIANCES_SUCCESS,
  type: CALL_API,
});
