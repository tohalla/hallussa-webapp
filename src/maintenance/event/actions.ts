import { map, pick }Â from "ramda";

import { CALL_API } from "../../store/middleware/api/actions";
import { ReduxAPICall } from "../../store/middleware/api/api";

export const FETCH_MAINTENANCE_EVENT_SUCCESS = "FETCH_MAINTENANCE_EVENT_SUCCESS";
// export const UPDATE_MAINTENANCE_EVENT_SUCCESS = "UPDATE_MAINTENANCE_EVENT_SUCCESS";
// export const DELETE_MAINTENANCE_EVENT_SUCCESS = "DELETE_MAINTENANCE_EVENT_SUCCESS";

export interface MaintenanceEventPayload {
  createdAt: string;
  description: string;
  resolvedAt: string;
  assignedTo: number;
}

export interface MaintenanceEventAction {
  type: string;
  payload: MaintenanceEventPayload;
}

export const fetchApplianceEvents = (
  organisation: number,
  appliance: number,
  {bypassCache = false} = {}
): ReduxAPICall => ({
  attemptToFetchFromStore: (state) => {
    if (!bypassCache && Array.isArray(state.entities.appliances[appliance].maintenanceEvents)) {
      return pick(
        map(String, state.entities.appliances[appliance].maintenanceEvents || []),
        state.entities.maintenanceEvents
      );
    }
    return undefined;
  },
  endpoint: `/organisations/${organisation}/appliances/${appliance}/maintenance-events`,
  method: "get",
  successType: FETCH_MAINTENANCE_EVENT_SUCCESS,
  type: CALL_API,
});
