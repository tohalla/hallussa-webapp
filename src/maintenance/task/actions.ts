import { MaintainerPayload } from "../../maintainer/actions";
import { CALL_API } from "../../store/middleware/api/actions";
import { ReduxAPICall } from "../../store/middleware/api/api";

export const FETCH_MAINTENANCE_TASK_SUCCESS = "FETCH_MAINTENANCE_TASK_SUCCESS";
// export const UPDATE_MAINTENANCE_TASK_SUCCESS = "UPDATE_MAINTENANCE_TASK_SUCCESS";
// export const DELETE_MAINTENANCE_TASK_SUCCESS = "DELETE_MAINTENANCE_TASK_SUCCESS";

export interface MaintenanceTaskPayload {
  updatedAt?: string;
  createdAt?: string;
  hash?: string;
  maintenanceEvent?: number;
  description?: string;
  maintainer?: number;
}

export interface MaintenanceTaskAction {
  type: string;
  payload: MaintenanceTaskPayload;
}

export const fetchMaintainerTasks = (
  maintainer: MaintainerPayload
): ReduxAPICall => ({
  endpoint: `/organisations/${maintainer.organisation}/maintainers/${maintainer.id}/maintenance-tasks`,
  method: "get",
  successType: FETCH_MAINTENANCE_TASK_SUCCESS,
  type: CALL_API,
});
