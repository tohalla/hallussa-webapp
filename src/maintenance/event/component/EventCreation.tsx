import classnames from "classnames";
import React, { useState } from "react";
import { useTranslation } from "react-i18next";

import { AppliancePayload } from "../../../appliance/actions";
import Button from "../../../component/button/Button";
import Restricted from "../../../component/Restricted";
import { contentHorizontalSpacing, rowContainer } from "../../../style/container";
import ScheduledMaintenanceForm from "./ScheduledMaintenanceForm";

interface Props {
  appliance: AppliancePayload;
}

export default ({appliance}: Props) => {
  const [activeForm, setForm] = useState();
  const {t} = useTranslation();

  return (
    <Restricted userRole={{allowManageMaintenanceTask: true}}>
      <h3>{t("maintenanceEvent.create.title")}</h3>
      {
        activeForm === "scheduled" ?
          <ScheduledMaintenanceForm appliance={appliance}/>
        :
          <div className={classnames(rowContainer, contentHorizontalSpacing)}>
            <Button onClick={setForm.bind(undefined, "scheduled")} plain={true}>
              {t("maintenanceEvent.create.scheduled.title")}
            </Button>
            <Button onClick={setForm.bind(undefined, "repetitive")} plain={true}>
              {t("maintenanceEvent.create.repetitive.title")}
            </Button>
          </div>
      }
    </Restricted>
  );
};
