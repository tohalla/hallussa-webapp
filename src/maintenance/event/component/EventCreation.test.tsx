import React from "react";

import { render } from "../../../__mocks__/runtime";
import { AppliancePayload } from "../../../appliance/actions";
import EventCreation from "./EventCreation";

describe("Maintenance EventCreation component", () => {
  const mockAppliance: AppliancePayload = {
    createdAt: new Date().toISOString(),
    description: "",
    hash: "_",
    id: 0,
    maintainers: [],
    name: "mock appliance",
    organisation: 0,
  };

  it("should render the component", () => {
    const {container} = render(
      <EventCreation appliance={mockAppliance}/>
    );

    expect(container).toMatchSnapshot();
  });
});
