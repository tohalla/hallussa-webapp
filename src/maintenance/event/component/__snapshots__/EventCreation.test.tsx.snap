// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Maintenance EventCreation component should render the component 1`] = `
<div>
  <h3>
    maintenanceEvent.create.title
  </h3>
  <div
    class="css-1fxb6ti css-1o5xcg5"
  >
    <button
      class="css-bmo2k2"
      type="button"
    >
      maintenanceEvent.create.scheduled.title
    </button>
    <button
      class="css-bmo2k2"
      type="button"
    >
      maintenanceEvent.create.repetitive.title
    </button>
  </div>
</div>
`;
